### PR TITLE
Improve the trivy CVE scanning output

### PR DESCRIPTION
**Pull request by hansinikarunarathne**
# Pull Request Template for Kubeflow manifests Issues

- Please include a summary of changes and the related issue. 
- List any dependencies that are required for this change. 
- Please delete the options that are not relevant.
- The following checklist will help you to satisfy the requirements.

## ✅ Contributor checklist
  - [X] All the commits have been _signed-off_  (To pass the `DCO` check)
  - [X] Submit the [Contributor License Agreements](https://cla.developers.google.com/clas) (To pass the `cla/google` check)

---     
 
> You can join the CNCF Slack and access our meetings at the [Kubeflow Community](https://www.kubeflow.org/docs/about/community/) website. Our channel on the CNCF Slack is here [**#kubeflow-platform**](https://app.slack.com/client/T08PSQ7BQ/C073W572LA2).

### CODE REVIEW

**Comment by juliusvonkohout** [comment]
Please also get rid of the /documents folder in this PR and address all comments from the previous PR.

**Comment by hansinikarunarathne** [comment]
> Please also get rid of the /documents AND /image_lists folderS in this PR and address all comments from the previous PR.

I fixed it. Can you check it ?

**Comment by hansinikarunarathne** [comment]
> I see for example
> 
> File ../image_lists/kf_latest_model-registry_images.txt successfully created File ../image_lists/kf_latest_all_images.txt successfully created Started scanning images Scanning images in ../image_lists/kf_latest_workbenches_images.txt Scanning centraldashboard:v1.9.0-rc.0
> 
> Scanning jupyter-web-app:v1.9.0-rc.0
> 
> ...
> 
> can we have something like
> 
> Scanning images in ../image_lists/kf_latest_workbenches_images.txt Scanning centraldashboard:v1.9.0-rc.0
> 
> Critical	High	Medium	Low
> 1	2	3	4
> Scanning jupyter-web-app:v1.9.0-rc.0
> 
> Critical	High	Medium	Low
> 5	6	7	8
> ...
> 
> ? So a mini table per image in the logs. This makes it easier to verify the numbers for third parties.

Added the table for reach log

File: hack/trivy_scan.py [hunk 1]
```diff
@@ -54,7 +55,7 @@ def save_images(wg, images, version):
     output_file = f"../image_lists/kf_{version}_{wg}_images.txt"
     with open(output_file, 'w') as f:
         f.write('\n'.join(images))
-    print(f"File {output_file} successfully created")
+    print(f"File {output_file} successfully created", flush=True)
```
Comment (AndersBennedsgaard) [inline hack/trivy_scan.py:58]: Wouldn't it make sense to create a `log` function (or name something else) which is just `print(..., flush=True)`?

File: hack/trivy_scan.py [hunk 2]
```diff
@@ -136,33 +136,34 @@ def extract_images(version):
...
-            subprocess.run(["trivy", "image", "--format", "json", "--output", scan_output_file, line], check=True)
+            result = subprocess.run(["trivy", "image", "--format", "json", "--output", scan_output_file, line], check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
+            print(result.stdout, flush=True)
 
             with open(scan_output_file, 'r') as json_file:
                 scan_data = json.load(json_file)
 
             if not scan_data.get('Results'):
-                print(f"No vulnerabilities found in {image_name}:{image_tag}")
+                print(f"No vulnerabilitiy found in {image_name}:{image_tag}", flush=True)
```
Comment (AndersBennedsgaard) [inline hack/trivy_scan.py:154]:
```suggestion
                print(f"No vulnerabilities found in {image_name}:{image_tag}", flush=True)
```

File: hack/trivy_scan.py [hunk 3]
```diff
@@ -136,33 +136,34 @@ def extract_images(version):
...
             else:
-                vulnerabilities = [
+                vulnerabilitiy = [
```
Comment (AndersBennedsgaard) [inline hack/trivy_scan.py:156]: Why rename this?
```suggestion
                vulnerabilities = [
```

File: hack/trivy_scan.py [hunk 4]
```diff
@@ -305,14 +317,14 @@ def extract_images(version):
...
 output_file = SUMMARY_OF_SEVERITY_COUNTS + '/summary_of_severity_counts_for_WGs_in_table.txt'
 with open(output_file, 'w') as f:
     f.write(str(table))
 
-print("Output saved to:", output_file)
-print("Severirty counts with images respect to WGs are saved in the" ,ALL_SEVERITY_COUNTS)
-print("Scanned Json reports on images are saved in" ,SCAN_REPORTS_DIR)
\ No newline at end of file
+print("Output saved to:", output_file, flush=True)
+print("Severirty counts with images respect to WGs are saved in the" ,ALL_SEVERITY_COUNTS, flush=True)
```
Comment (AndersBennedsgaard) [inline hack/trivy_scan.py:329]:
```suggestion
print("Severity counts with images respect to WGs are saved in the", ALL_SEVERITY_COUNTS, flush=True)
```

File: .github/workflows/trivy.yaml [hunk 5]
```diff
@@ -38,8 +38,8 @@ jobs:
         run: |
           pip install prettytable
 
-      # Run the shell script (trivy_scan.py)
+      # Run the python script (trivy_scan.py)
```
Comment (AndersBennedsgaard) [inline .github/workflows/trivy.yaml:41]: Redundant comment
```suggestion
```

File: hack/trivy_scan.py [hunk 6]
```diff
@@ -148,19 +151,19 @@ def extract_images(version):
...
                     if 'Vulnerabilities' in result and result['Vulnerabilities']
                 ]
 
-                if not vulnerabilities:
+                if not vulnerabilitiy:
                     print(f"The vulnerability detection may be insufficient because security updates are not provided for {image_name}:{image_tag}", flush=True)
                 else:
                     severity_counts = {"LOW": 0, "MEDIUM": 0, "HIGH": 0, "CRITICAL": 0}
-                    for vulns in vulnerabilities:
+                    for vulns in vulnerabilitiy:
```
Comment (juliusvonkohout) [inline hack/trivy_scan.py:166]: "for vulnerability in vulnerabilities:" could make more sense. Or is the highest level a vulnerabilities_map ?

File: .github/workflows/trivy.yaml [hunk 7]
```diff
@@ -43,10 +43,3 @@ jobs:
         run: |
           cd hack
           python3 trivy_scan.py
-
```
Comment (juliusvonkohout) [inline .github/workflows/trivy.yaml]: Just leave it commented out, but do not delete


      # Upload the artifact
      - name: Upload trivy scanned_results
        uses: actions/upload-artifact@v4
        with:
          name: trivy_scanned_results
          path: image_lists/

**Comment by juliusvonkohout** [comment]
ah and it should not run on pull request by default only on the master branch. This blocks other PRs such as https://github.com/kubeflow/manifests/pull/2769

**Comment by juliusvonkohout** [comment]
@AndersBennedsgaard I hope you can verify the CVE numbers per image and WG independently.

**Comment by AndersBennedsgaard** [comment]
> @AndersBennedsgaard I hope you can verify the CVE numbers per image and WG independently.

@juliusvonkohout did you mean to tag @hansinikarunarathne instead ?

**Comment by juliusvonkohout** [comment]
> > @AndersBennedsgaard I hope you can verify the CVE numbers per image and WG independently.
> 
> @juliusvonkohout did you mean to tag @hansinikarunarathne instead ?

No, i would like others to validate as well.

**Comment by AndersBennedsgaard** [comment]
@juliusvonkohout I'm not entirely sure what you mean. Do you want me to check if all Kubeflow images are caught by the script, or do you want me to run Trivy on the same images? Running Trivy to verify the numbers are the same when running locally version in GH actions seems redundant - especially due to the time it takes to scan all the images.

However, speaking of the time to scan: you could probably introduce threading/multi-processing when scanning different images to speed up the script by a lot, since it should be possible to run the `trivy` command concurrently.

**Comment by juliusvonkohout** [comment]
> Do you want me to check if all Kubeflow images are caught by the script

Yes, if you think that the CVE numbers are not worth checking.

File: hack/trivy_scan.py (hunk 1 above)
Comment (hansinikarunarathne) [reply hack/trivy_scan.py]: I made the change

File: hack/trivy_scan.py (hunk 4 above)
Comment (hansinikarunarathne) [reply hack/trivy_scan.py]: > made the change

File: .github/workflows/trivy.yaml (hunk 5 above)
Comment (hansinikarunarathne) [reply .github/workflows/trivy.yaml]: Removed this line

File: hack/trivy_scan.py (hunk 2 above)
Comment (hansinikarunarathne) [reply hack/trivy_scan.py]: Fixed it

File: hack/trivy_scan.py (hunk 3 above)
Comment (hansinikarunarathne) [reply hack/trivy_scan.py]: fixed it

File: hack/trivy_scan.py [hunk 8]
```diff
@@ -148,19 +151,19 @@ def extract_images(version):
...
             else:
-                vulnerabilities = [
+                vulnerabilitiy = [
                     result['Vulnerabilities']
                     for result in scan_data['Results']
                     if 'Vulnerabilities' in result and result['Vulnerabilities']
                 ]
 
-                if not vulnerabilities:
+                if not vulnerabilitiy:
```
[the remark this answers is not in the document]
Comment (hansinikarunarathne) [reply hack/trivy_scan.py]: Chnaged it

**Comment by hansinikarunarathne** [comment]
> ah and it should not run on pull requests by default. It should run only if a PR has been merged into the master branch. This blocks other PRs such as #2769 because it takes 40 minutes to run.
> 
> Add i think the formatting is a bit off It should look like
> 
> ```
> Started scanning images
> Scanning images in ../image_lists/kf_latest_workbenches_images.txt
> Scanning  docker.io/kubeflownotebookswg/centraldashboard:v1.9.0-rc.0
> +----------+------+--------+-----+
> | Critical | High | Medium | Low |
> +----------+------+--------+-----+
> |    1     |  11  |   37   |  4  |
> +----------+------+--------+-----+
> 
> Scanning  docker.io/kubeflownotebookswg/jupyter-web-app:v1.9.0-rc.0
> +----------+------+--------+-----+
> | Critical | High | Medium | Low |
> +----------+------+--------+-----+
> |    1     |  11  |   39   |  67 |
> +----------+------+--------+-----+
> ```
> 
> instead of
> 
> ```
> Started scanning images
> Scanning images in ../image_lists/kf_latest_workbenches_images.txt
> Scanning  docker.io/kubeflownotebookswg/centraldashboard:v1.9.0-rc.0
> 
> +----------+------+--------+-----+
> | Critical | High | Medium | Low |
> +----------+------+--------+-----+
> |    1     |  11  |   37   |  4  |
> +----------+------+--------+-----+
> Scanning  docker.io/kubeflownotebookswg/jupyter-web-app:v1.9.0-rc.0
> 
> +----------+------+--------+-----+
> | Critical | High | Medium | Low |
> +----------+------+--------+-----+
> |    1     |  11  |   39   |  67 |
> +----------+------+--------+-----+
> ```

I made the adjustment according to this

**Review by AndersBennedsgaard — APPROVED** [review]
LGTM

File: hack/trivy_scan.py [hunk 9]
```diff
@@ -80,14 +86,13 @@ def extract_images(version):
...
-                            kbuild = result.stdout
                         except subprocess.CalledProcessError as e:
-                            print(f"ERROR:\t Failed \"kustomize build\" command for directory: {root}. See error above")
+                            log(f"ERROR:\t Failed \"kustomize build\" command for directory: {root}. See error above")
                             continue
                         
                         # Use regex to find lines with 'image: <image-name>:<version>' or 'image: <image-name>'
                         # and '- image: <image-name>:<version>' but avoid environment variables
-                        kimages = re.findall(r'^\s*-?\s*image:\s*([^$\s:]+(?:\:[^\s]+)?)$', kbuild, re.MULTILINE)
+                        kimages = re.findall(r'^\s*-?\s*image:\s*([^$\s:]+(?:\:[^\s]+)?)$', result.stdout, re.MULTILINE)
```
Comment (juliusvonkohout) [inline hack/trivy_scan.py:95]: Please use clear variable names, so either images or kustomize_images.

**Review by juliusvonkohout — CHANGES_REQUESTED** [review]

File: hack/trivy_scan.py (hunk 9 above)
Comment (hansinikarunarathne) [reply hack/trivy_scan.py]: > Did the change

**Review by juliusvonkohout — APPROVED** [review]

**Pull request merged**